### DISCUSSION TITLE
RP2040 - SD Card Issue

### DISCUSSION THREAD

**Discussion by DoubleOut** [STELLA-Landsat/STELLA#61] [Build Notes: How do I make a STELLA?]
Greetings! I'm building a Stella-AQ but I'm having issues with the Thing Plus RP2040 mounting the SD card. The error is "OSError: no SD card". I know I'm missing something basic.

Trouble shooting steps I've tried with both the recommended card (45MB) as well as a spare (4G) card I have:
* Confirmed read and write using a card reader on my computer
* Formatted (re-formatted) to FAT
* Purchased a new board to ensure it wasn't a bad RP2040 board
* With and without the additional components attached
* Previous and next versions of Circuit python: 9.0.3, 9.0.4, 9.2(beta)

Both the AQ code and the code below generate the same error. I've tinkered a bit with the SPI settings, but since this doesn't seem to be a common problem, it's probably something simple.

Any insight would be appreciated!

```
import os
import board
import sdcardio
import storage

sd = sdcardio.SDCard(board.SPI(), board.SD_CS)  # error on this line
vfs = storage.VfsFat(sd)
storage.mount(vfs, '/sd')
os.listdir('/sd')
```

**Comment by paulmirel** [comment]
Hello! Thanks for your interest and enthusiasm for STELLA! 

There's a change in how SD card mounting is done from CircuitPython 8 to CircuitPython 9. We haven't upgraded our codes to version 9 yet. 

Please try loading CircuitPython 8 and using the matching libraries. I expect that will work for you. Please let us know!

**Reply by DoubleOut** [reply]
Thank you for the quick response. I've ruled out any issues with the STELLA code. The error is on a CircuitPython built in library `sdcardio`. I get the same error with CircuitPython v8.1 and v8.2.7, both cards and both boards.

Today I will purchase some additional cards since they aren't expensive. I may also my try configuring SPI manually but that shouldn't be necessary on newer boards. 

This is the code I'm using:
```
import board
import sdcardio
sd = sdcardio.SDCard(board.SPI(), board.SD_CS)
```

**Comment by paulmirel** [comment]
I think you'll need to change to using sdioio, which is more current than sdcardio. 

import sdioio
sdcard = sdioio.SDCard(
        clock=board.SDIO_CLOCK,
        command=board.SDIO_COMMAND,
        data=board.SDIO_DATA,
        frequency=25000000)

The test code for the SD card looks for sdioio first, and, failing that, looks for sdcardio. It may be that you need to be using sdioio. I've included that test code here. 

[05. SD card.zip](https://github.com/STELLA-Landsat/STELLA/files/15370411/05.SD.card.zip)

Other possibles: 

We've had problems with the 512MB SD cards. They would sometimes fail to complete a write command, causing the code to hang up. Replacing them with a name-brand card solved the problem. 

It's possible, though unlikely, that the particular Thing Plus you are using has a wiring fault somewhere on the SPI bus connection to the SD card. You can reduce that uncertainty by trying another Thing Plus of the the same type.

**Comment by paulmirel** [comment]
Having some formatting issues. the "sdcard =" line is all one statement. 

import sdioio
sdcard = sdioio.SDCard( clock=board.SDIO_CLOCK, command=board.SDIO_COMMAND, data=board.SDIO_DATA, frequency=25000000 )

**Comment by DoubleOut** [comment]
Thank you @paulmirel , I am making progress here. The single line `sdcard =` code originated from some older AQ code. It appears that code was taken down last month so it may be undergoing revisions. The other clue I have is that SparkFun boards don't use the default SPI pin for the SD card. 

I'm working through the code you (and others in other forums) have posted. I will post here when I get this working!

**Comment by DoubleOut** [comment]
Thank you @paulmirel, I can confirm that `sdcardio` works well with the RP2040! 

Just curiosity, is the reason for including `sdioio` in the sample code for supporting other boards?

```
import board
import busio
import sdcardio
import storage

# Thing Plus RP2040 - CircuitPython 8.2.6
cs = board.SD_CS
sd_card_spi = busio.SPI(board.SD_SCK, MOSI=board.SD_MOSI, MISO=board.SD_MISO)
sdcard = sdcardio.SDCard(sd_card_spi, cs)
vfs = storage.VfsFat(sdcard)

# get the current working directory
print("The current working directory is: ", vfs.getcwd())

# list files on the sd card
for f in vfs.ilistdir():
    print(f)
```
